From: SeongJae Park <sj@kernel.org>
Date: Sat, 14 Feb 2026 11:31:49 -0800
Subject: [PATCH] mm/damon/core: handle online-tuned min_nr_regions

To respect min_nr_regions, per-region size limit is set as the total
size of monitoring regions divided by the min_nr_regions.  It is updted
after ops update operation, because 'vaddr' updates the monitoring
regions boundaries in the operation.  That is, the total size of
monitoring regions can be changed in the operation, so the limit is
calculated again.  damon_call() callers can also update the regions
boundasry and min_nr_regions.  Foo...

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 86a7595b22eb7..e29a17e30c2d1 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2808,6 +2808,8 @@ static int kdamond_fn(void *data)
 		unsigned long next_aggregation_sis = ctx->next_aggregation_sis;
 		unsigned long next_ops_update_sis = ctx->next_ops_update_sis;
 		unsigned long sample_interval = ctx->attrs.sample_interval;
+		/* cache min_nr_regions; it can be changed in damon_call() */
+		unsigned long old_min_nr_regions = ctx->attrs.min_nr_regions;
 
 		if (kdamond_wait_activation(ctx))
 			break;
@@ -2883,6 +2885,10 @@ static int kdamond_fn(void *data)
 				ctx->ops.update(ctx);
 			sz_limit = kdamond_get_apply_max_region_sz(ctx);
 		}
+
+		/* damon_call() callers could updated min_nr_regions */
+		if (ctx->attrs.min_nr_regions != old_min_nr_regions)
+			sz_limit = kdamond_get_apply_max_region_sz(ctx);
 	}
 done:
 	damon_destroy_targets(ctx);
-- 
2.47.3

